desc/magnetic_fields.py

```diff
@@ -1819,9 +1836,875 @@ def from_surface(
+            surface current on the given winding surface
+        """
+        surface_current_field = self
+        nfp = surface_current_field.Phi_basis.NFP
+
+        net_toroidal_current = surface_current_field.I
+        net_poloidal_current = surface_current_field.G
+        assert not jnp.isclose(net_toroidal_current, 0) or not jnp.isclose(
+            net_poloidal_current, 0
+        ), (
+            "Detected both net toroidal and poloidal current are both zero, "
+            "this function cannot find windowpane coils"
+        )
+
+        winding_surf = surface_current_field
+
+        ################################################################
+        # find current helicity
+        ################################################################
+        # we know that I = -(G - G_ext) / (helicity * NFP)
+        # if net_toroidal_current is zero, then we have modular coils,
+        # and just make helicity zero
+        helicity = safediv(
+            -net_poloidal_current, net_toroidal_current * nfp, threshold=1e-8
+        )
+        phi_slope = jnp.sign(helicity)
+
+        theta_coil = jnp.linspace(0, 2 * jnp.pi, 128)
+        if not jnp.isclose(helicity, 0):
+            # helical coils, we need finer toroidal resolution
+            zeta_coil = jnp.linspace(
+                0, 2 * jnp.pi / nfp, round(128 * jnp.abs(helicity))
+            )
+            zetal_coil = zeta_coil
+        else:
+            # modular coils
+            theta_coil = jnp.linspace(0, 2 * jnp.pi, 128)
+            zeta_coil = jnp.linspace(0, 2 * jnp.pi / nfp, 128)
+            dz = zeta_coil[1] - zeta_coil[0]
+            zetal_coil = jnp.arange(-jnp.pi / nfp, (2 + 1 / nfp) * jnp.pi, dz)
+
+        ################################################################
+        # find contours of constant phi
+        ################################################################
+        # make linspace contours
+        if not jnp.isclose(helicity, 0):
+            # helical coils
+            # we start them on zeta=0 plane, so we will find contours
+            # going up from 0 to I (corresponding to zeta=0, and theta increasing)
+            contours = jnp.linspace(
+                0, jnp.abs(net_toroidal_current), desirednumcoils + 1, endpoint=True
+            )
+            contours = jnp.sort(jnp.sign(phi_slope) * jnp.asarray(contours))
+            coil_current = jnp.abs(net_toroidal_current) / desirednumcoils
+
+        else:
+            # modular coils
+            # go from zero to G
+            contours = jnp.linspace(
+                0, jnp.abs(net_poloidal_current), desirednumcoils + 1, endpoint=True
+            ) * jnp.sign(net_poloidal_current)
+            contours = jnp.sort(jnp.asarray(contours))
+            coil_current = net_poloidal_current / desirednumcoils
+
+        # TODO: change this so that  this we only need Ncoils length array
+
+        theta_full = theta_coil
+        if not jnp.isclose(helicity, 0):
```

Comment:
sure, so we want a grid that we can evaluate Phi on to find contours of constant Phi to turn into coils

we really only care about things from phi = 0 to 2pi/NFP, and theta=0 to 2pi, and we will then follow a certain amount of constant Phi contours to find our coils. BUT, if the contour say starts at zeta=0, theta=pi and then goes to zeta=2pi/NFP, theta=3pi, if we only are plotting theta in [0,2pi], we will lose half  of this contour. So, we want to increase the theta range so that we can catch these coils.

that is for helical, for modular that is not a big issue but we still want to put some leeway around zeta=0 and zeta=2pi (should be 2pi/NFP but I need to modify that) so that any contours which may be  a bit wiggly around zeta=0 and zeta=2pi are fully caught by the contour finding